subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -95,9 +101,40 @@ public void execute(@Nonnull String testClassName) {
 
         private void processAllTestClasses() {
             Launcher launcher = LauncherFactory.create();
-            launcher.registerTestExecutionListeners(new JUnitPlatformTestExecutionListener(resultProcessor, clock, idGenerator));
-            launcher.execute(createLauncherDiscoveryRequest(testClasses));
+            TestExecutionListener listener = new JUnitPlatformTestExecutionListener(resultProcessor, clock, idGenerator);
+            LauncherDiscoveryRequest discoveryRequest = createLauncherDiscoveryRequest(testClasses);
+            launcher.registerTestExecutionListeners(listener);
+
+            if (spec.isDryRun()) {
+                TestPlan testPlan = launcher.discover(discoveryRequest);
+                executeDryRun(testPlan, listener);
+            } else {
+                launcher.execute(createLauncherDiscoveryRequest(testClasses));
+            }
+        }
+    }
+
+    private void executeDryRun(TestPlan testPlan, TestExecutionListener listener) {
+        List<TestIdentifier> testIdentifiers = new LinkedList<>();
+
+        for (TestIdentifier root : testPlan.getRoots()) {
+            testIdentifiers.add(root);
+            testIdentifiers.addAll(testPlan.getDescendants(root));
         }
+
+        listener.testPlanExecutionStarted(testPlan);
+
+        for (TestIdentifier identifier : testIdentifiers) {
+            listener.executionStarted(identifier);
+        }
+
+        Collections.reverse(testIdentifiers);
```

Comment:
I think it's needed it because i'm "flatted" graph of testIdentifiers to linkedList of them